src/state/block_tree.rs

```diff
@@ -0,0 +1,565 @@
+        for block in self.blocks_in_branch(*root) {
+            wb.delete_children(&block);
+            wb.delete_pending_app_state_updates(&block);
+            wb.delete_block_validator_set_updates(&block);
+
+            if let Ok(Some(data_len)) = self.block_data_len(&block) {
+                wb.delete_block(&block, data_len)
+            }
+        }
+    }
+
+    /// Performs depth-first search to collect all blocks in a branch into a single iterator.
+    pub fn blocks_in_branch(&self, root: CryptoHash) -> impl Iterator<Item = CryptoHash> {
+        let mut stack: Vec<CryptoHash> = vec![root];
+        let mut branch: Vec<CryptoHash> = vec![];
+
+        while let Some(block) = stack.pop() {
+            if let Ok(children) = self.children(&block) {
+                for child in children.iter() {
+                    stack.push(*child)
+                }
+            };
+            branch.push(block)
+        }
+        branch.into_iter()
+    }
+
+    /* ↓↓↓ Extra state getters for convenience ↓↓↓ */
+
+    pub fn contains(&self, block: &CryptoHash) -> bool {
+        self.block(block).is_ok_and(|block_opt| block_opt.is_some())
+    }
+
+    pub(crate) fn highest_view_with_progress(&self) -> Result<ViewNumber, BlockTreeError> {
+        Ok(
+            max(
+                self.highest_view_entered()?,
+                max(
+                    self.highest_qc()?.view,
+                    self.highest_tc()?.map(|tc| tc.view).unwrap_or(ViewNumber::init()),
+                )
+            )
+        )
+    }
+
+    pub(crate) fn highest_committed_block_height(&self) -> Result<Option<BlockHeight>, BlockTreeError> {
+        let highest_committed_block = self.highest_committed_block()?;
+        if let Some(block) = highest_committed_block {
+            Ok(self.block_height(&block)?)
+        } else {
+            Ok(None)
+        }
+    }
+
+    /* ↓↓↓ WriteBatch commit ↓↓↓ */
+
+    pub fn write(&mut self, write_batch: BlockTreeWriteBatch<K::WriteBatch>) {
+        self.0.write(write_batch.0)
+    }
+
+    /* ↓↓↓ Snapshot ↓↓↓ */
+
+    pub fn snapshot(&self) -> BlockTreeSnapshot<K::Snapshot<'_>> {
+        BlockTreeSnapshot::new(self.0.snapshot())
+    }
+
+    /* ↓↓↓ Get AppBlockTreeView for ProposeBlockRequest and ValidateBlockRequest */
+
+    pub(crate) fn app_view<'a>(&'a self, parent: Option<&CryptoHash>) -> Result<AppBlockTreeView<'a, K>, BlockTreeError> {
+        let highest_committed_block_height = self.highest_committed_block_height()?;
+        let parent = 
+            match parent {
+                None => None,
+                Some(&b) => Some(b)
+            };
+
+        // Obtain an iterator over the ancestors starting from the parent, all the way until genesis, 
+        // from newest (parent) to oldest.
+        let ancestors_iter =
+            successors(
+            parent, 
+            |b| 
+                self.block_justify(b).ok()
```

Comment:
This `.ok()` quietly turns the Result returned from `block_justify` into an Option. Did you consider any better ways to handle a `BlockTreeError` in the evaluation of this closure? E.g., maybe having the iterator be over `Result<CryptoHash, BlockTreeError>` and panicking when the iterator is consumed if we come across a `BlockTreeError`?

Reply:
My intention was that the iterator stops once the second argument to successors returns `None`. In our case, once a block with no `justify` is encountered, we'd want to stop iterating. On the other hand, we stop once we see a genesis QC, so the former is not necessary. 

But I think you're right, we shouldn't be using `ok()` here. Returning an error would also make sense, though I'd rather not panic here. The problem is that we are in a closure so we can't propagate the error from the closure. I'll think about how to fix this, but if you have an idea (seems like you do but I am not sure if I understand correctly) feel free to commit this change?

Reply:
Extracted into issue #40.